classes/SystemOverwrites/extAbstractFunction.sc

```diff
@@ -0,0 +1,7 @@
++ AbstractFunction {
+
+	// TODO: migrate thresh2 methods to SC common lib
+	thresh2 { |function|
+		^this * (this.abs >= function).asInteger
```

Comment:
`.asInteger` will break for UGens. The function can be a UGen that changes over time. If you want to guarantee that a function returning an integer returns an integer, we have to find another way (not sure yet what would be the best way).

Btw for math operations, it is usually much better to use integer floats than `Integer`, because the former are 64 bit. Is theer a reason you need to keep ints?

Reply:
> .asInteger will break for UGens (and other functions). 

Wow, I have no idea how this was working before. I _must_ have recompiled after making the help examples, etc., but there it is...

Anyway, I originally didn't use `.asInteger`, which worked fine for UGens, but then `Thunk`, a subclass of `AbstractFunction`, broke:
```supercollider
(
a = Thunk({ rrand(-2.0, 2.0) });
// t = a.thresh(3); // works
t = a.thresh2(3);   // breaks
t.value
)
>> ERROR: binary operator '*' failed.
RECEIVER:
   false
ARGS:
   Float -0.340691   00000000 BFD5CDE2
   nil
```
Should this be expected to work as it's written? `thresh` and other binary ops work in this test.

> Btw for math operations, it is usually much better to use integer floats than Integer, because the former are 64 bit. Is theer a reason you need to keep ints?

I typically is `.asInteger` when converting Booleans to do math with them. Is there a better way?

Reply:
Ah, yes. I'm sorry. Here is a solution that works for me:

```supercollider
+ AbstractFunction {

	thresh2 { |function, adverb|
		^this.composeBinaryOp('thresh2', function, adverb)
	}

}

+ SimpleNumber {

	thresh2 { |thresh, adverb|
		^if(this.abs >= thresh) { this } { thresh }
	}
}

+ UGen {
	thresh2 { |function, adverb|
		^this * (this.abs >= function)
	}
}
```

Reply:
Thanks for the update. `Thunk` now works.
Your suggested version of `SimpleNumber:-thresh2` actually clips the value rather than forcing it to 0. Your original objection to using `asInteger` was that it would break with UGens, but now that we have a separate method for `UGen`, are you OK with `SimpleNumber:-thresh2` as it is? It currently works to return the same type as the receiver. To change the method and maintain the type returned would seem to need separate methods for `Float` and `Integer`.